### DISCUSSION TITLE
Cannot inject a custom principal into unauthenticated request

### DISCUSSION THREAD

**Discussion by matthias-mueller** [dropwizard/dropwizard#5267] [General]
When a method in a Jersey resource has a parameter (for instance of some type `User`) annotated with `@Auth` and authentication fails, the following exception is thrown:

```
ERROR [2022-05-02 08:34:03,534] io.dropwizard.jersey.errors.IllegalStateExceptionMapper: Error handling a request: 1fa109dc501e3397
! java.lang.IllegalStateException: Cannot inject a custom principal into unauthenticated request
! at io.dropwizard.auth.PrincipalContainerRequestValueFactory.provide(PrincipalContainerRequestValueFactory.java:25)
! at io.dropwizard.auth.AuthValueFactoryProvider.lambda$createValueProvider$1(AuthValueFactoryProvider.java:51)
! at org.glassfish.jersey.server.spi.internal.ParamValueFactoryWithSource.apply(ParamValueFactoryWithSource.java:50)
```

While I can clearly see where this is coming from, I am not sure that it is always practical. It seems that Dropwizard requires authentication prior to injecting a principal, which is somewhat reasonable for many scenarios. For a particular application, I need a resource that also accepts anonymous (e.g. `null`) users to serve public content (whereas authenticated and authorized users see the public content plus X with X being the protected content).

As a workaround, I could append an AnonymousAuthenticator at the end of an Auth Chain and let this one create the `User` object. If Chained authentication is not necessary, a separate code path in a dedicated authenticator might also do the job (but is a dead end as soon as the application requires an additional authentication mechanism.)

So in general I can mitigate my particular problem in some ways but none of them feels 100% right. Is there a good reason why Dropwizard does not simply inject a NULL Principal if authentication cannot be performed?

**Comment by zUniQueX** [comment]
@matthias-mueller You can use an `@Auth Optional<T>` principal for your endpoint. This will let unauthenticated users pass. But you can check, if the principal is present, and add custom logic for this case.

**Comment by matthias-mueller** [comment]
Thanks for the hint @zUniQueX - While I can confirm that `@Auth Optional<T>` is working, I see some general issues with the control flow and the interplay of `@Auth`, `@PermitAll`, `@RolesAllowed` and `Optional<Principal>`. I have created four simple cases below.

__Resource Class__
```java
@Path("/test/example")
public class ExampleResource {

    private static final Logger LOGGER = LoggerFactory.getLogger(ExampleResource.class);

    public ExampleResource() {
    }

    @GET
    @Path("/optional-user")
    public String getOptionalUser(
            @Context final UriInfo requestUri, @Auth final Optional<JWTUser> user
    ) {
        LOGGER.info("found user {}", user);
        return Objects.toString(user);
    }

    @GET
    @Path("/nullable-user")
    public String getOptionalUser(
            @Context final UriInfo requestUri, @Auth final JWTUser user
    ) {
        LOGGER.info("found user {}", user);
        return Objects.toString(user);
    }

    @GET
    @Path("/optional-user-with-roles-allowed")
    @PermitAll
    public String getOptionalUserWithRolesAllowed(
            @Context final UriInfo requestUri, @Auth final Optional<JWTUser> user
    ) {
        LOGGER.info("found user {}", user);
        return Objects.toString(user);
    }

    @GET
    @Path("/nullable-user-with-roles-allowed")
    @PermitAll
    public String getNullableUserWithRolesAllowed(
            @Context final UriInfo requestUri, @Auth final JWTUser user
    ) {
        LOGGER.info("found user {}", user);
        return Objects.toString(user);
    }

}

```

__Test Class__
```java
@ExtendWith(DropwizardExtensionsSupport.class)
public final class RestServerExampleTest extends RestServerTestAbstract {


    private static final Logger LOGGER = LoggerFactory.getLogger(RestServerExampleTest.class);

    private static final DropwizardAppExtension<PermissionsRestServerConfiguration> EXT;

    static {
        LOGGER.info(ready);
        EXT = new DropwizardAppExtension<>(PermissionsRestServer.class, ResourceHelpers.resourceFilePath("permissions.yml"), updateDbConfig());
    }

    @Test
    public void getOptionalUser() {
        RestAssured.given()
                .contentType(MediaType.APPLICATION_JSON)
                .when()
                .get("/test/example/optional-user")
                .then()
                .statusCode(HttpStatus.OK_200);
    }

    @Test
    public void getNullableUser() {
        RestAssured.given()
                .contentType(MediaType.APPLICATION_JSON)
                .when()
                .get("/test/example/nullable-user")
                .then()
                .statusCode(HttpStatus.OK_200);
    }

    @Test
    public void getOptionalUserWithRolesAllowed() {
        RestAssured.given()
                .contentType(MediaType.APPLICATION_JSON)
                .when()
                .get("/test/example/optional-user-with-roles-allowed")
                .then()
                .statusCode(HttpStatus.UNAUTHORIZED_401);
    }

    @Test
    public void getNullableUserWithRolesAllowed() {
        RestAssured.given()
                .contentType(MediaType.APPLICATION_JSON)
                .when()
                .get("/test/example/nullable-user-with-roles-allowed")
                .then()
                .statusCode(HttpStatus.UNAUTHORIZED_401);
    }

}
```

__Test Results__
* getOptionalUser - __200 OK__
* getNullableUser - __500__ 
* getOptionalUserWithRolesAllowed - __200__ 
* getNullableUserWithRolesAllowed - __500__ 

__Observations__
> getOptionalUser
* works as expected

> getNullableUser 
* "Cannot inject a custom principal into unauthenticated request"
* this error is misleading because DW can actually inject a custom principal into an unauthenticated request (see `getOptionalUser`). It just refuses to do that if the target type is not an `Optional<T>`. At least there is no technical reason why DW cannot not inject NULL values if no principal could be obtained.
* Personal comment: That decision is at least arguable, because it forces people to use `<Optional>` in their APIs

> getNullableUserWithRolesAllowed
* "Cannot inject a custom principal into unauthenticated request"
* There should not be any injection attempt at all because `@PermitAll` already forbids access by unauthenticated users
* DW should return __401__-Unauthorized. The inability to authenticate a user is already a sufficient condition for the `@PermitAll` check and any code inside `@PermitAll getOptionalUserWithRolesAllowed()` would never be executed within that request

> getOptionalUserWithRolesAllowed
* The user could not be authenticated but is actually granted access - I would not expect this to happen with `@PermitAll`

__@RolesAllowed__
* If `@PermitAll` is substituted with `@RolesAllowed({'A'})` code `403 - Forbidden` is returned
* `401 - Unauthorized` - request has not been applied because it lacks valid authentication credentials for the target resource
* `403 - Forbidden` - the server understood the request but refuses to authorize it...If authentication credentials were provided in the request, the server considers them insufficient to  grant access
* Source: https://www.freecodecamp.org/news/http-401-error-vs-http-403-error-status-code-responses-explained/
* So if DW distinguishes between `401` and `403` returning a `401` seems more appropriate if no principal could be obtained

(Edited)

p.s.: https://stackoverflow.com/a/43393182
> The `@PermitAll`, `@RolesAllowed`, etc are only ways of Jersey handling authorization. It is expected that there already be an authenticated user. How Jersey checks this is by getting the principal from the servlet request. If there is no principal, then it is assumed that there is not authenticated user, and then will not allow anyone through, even if you have @PermitAll, __as even @PermitAll requires a user to be authenticated.__

**Comment by matthias-mueller** [comment]
(had to edit the above examples due to some code errors)

**Comment by zUniQueX** [comment]
@matthias-mueller Thanks for providing the test cases. Unfortunately, the `@RolesAllowed` things are a bit tricky. Jersey states in [its JavaDoc](https://eclipse-ee4j.github.io/jersey.github.io/apidocs/2.35/jersey/org/glassfish/jersey/server/filter/RolesAllowedDynamicFeature.html):
> If the &#64;PermitAll annotation is declared and is not overridden then this filter will not be applied.

[The implementation](https://github.com/eclipse-ee4j/jersey/blob/master/core-server/src/main/java/org/glassfish/jersey/server/filter/RolesAllowedDynamicFeature.java#L83) literally does nothing with this annotation. So the endpoints annotated with `@PermitAll` should behave the same as the corresponding non-annotated ones do (at least from Jersey's point of view).

The exception with the non-optional principal is a thing of definition in my opinion. If the message "Cannot inject a custom principal into unauthenticated request" is issued, an other definition than yours could be that a principal is defined as such only if it is _valid_, e.g. not `null`. Then the message is correct that no principal can be injected. Therefore the handling with the `Optional<T>` principal types is provided, as this allows the declaration that the principal may be unset.

**Comment by matthias-mueller** [comment]
On the 401 vs 403 return code question I have created a jersey issue: https://github.com/eclipse-ee4j/jersey/issues/5062 because their [JavaDoc](https://eclipse-ee4j.github.io/jersey.github.io/apidocs/2.35/jersey/org/glassfish/jersey/server/filter/RolesAllowedDynamicFeature.html) is ambiguous (The combination `403 (Not Authenticated)` is not defined in RFC7235)

**Comment by matthias-mueller** [comment]
@zUniQueX Thanks for the explanation! The [Jersey specification](https://download.oracle.com/otndocs/jcp/common_annotations-1_3-mrel3-spec/) on that annotation is not much clearer than the JavaDoc. It boils down to whether an unauthenticated user has a security role or not. (Which is not within the scope of this issue.)

It seems that [Spring](https://www.baeldung.com/spring-security-expressions#3-permitall-denyall) uses the same conventions as DW - so there is probably little surprise for developers (except for non Jersey experts like me).

**Comment by zUniQueX** [comment]
> On the 401 vs 403 return code question I have created a jersey issue: https://github.com/eclipse-ee4j/jersey/issues/5062 because their [JavaDoc](https://eclipse-ee4j.github.io/jersey.github.io/apidocs/2.35/jersey/org/glassfish/jersey/server/filter/RolesAllowedDynamicFeature.html) is ambiguous (The combination `403 (Not Authenticated)` is not defined in RFC7235)
> 

@matthias-mueller Thanks. As this is then a documentation problem on Jersey's side only, I'll close this issue, as we don't have to track their status on that topic here.

**Comment by matthias-mueller** [comment]
@zUniQueX Last part of my response is on whether to inject NULL into non-optional principals or not.

* `Optional` is primarily intended _as a return type_ (see Javadoc citation below)
> Optional is primarily intended for use as a method return type where there is a clear need to represent "no result," and where using null is likely to cause errors. A variable whose type is Optional should never itself be null; it should always point to an Optional instance.
* If something can be derived on `Optional`'s use and purpose _as an argument_ it is this part:
> ... where using null is likely to cause errors
* Since the decision of `Optional<Principal>` vs `nullable Pricipal` is up to DW users, it is an opt-in procedure users of `nullable Pricipal` should be aware of NULL principals
* Some users might not be aware of `Optional<Principal>` and use `nullable Pricipal` without care. NPE is still to the rescue and has the same effect as `IllegalStateException` at runtime. (And there`s [JEP 358](https://openjdk.java.net/jeps/358) which helps to debug those situations in JRE >=17)
* Another argument is symmetry / consistency in the framework: For Resource classes, DW gives users the freedom to opt into `Optional<T>` or `nullable T` for parameter injection. I don't see any reason why the behavior/treatment of `@Auth` should be different from `@Param` or `@UriInfo`

**Reply by matthias-mueller** [reply]
(p.s. This was the original issue which is still not solved)

**Reply by zUniQueX** [reply]
@matthias-mueller I wouldn't say this behavior is inconsistent. It mirrors the return value of the `Authenticator`. There you return a custom principal or `Optional.empty()`, if there is no principal. Therefore I think it is more consistent to have the same behavior in the resource method.

> NPE is still to the rescue and has the same effect as `IllegalStateException` at runtime.

This isn't completely correct. If users provide a parameter like `@Auth User user` in their resource method, they rely on an authenticated user. When this is the only purpose, then they wouldn't dereference the user at all and no NPE would be thrown. If a user only uses authentication and no authorization in an application, with your approach an explicit `null`-check would be necessary then.

**Reply by matthias-mueller** [reply]
I was talking about _consistent injection behavior_ compared to the other annotations (not symmetry with potential producers inside DW)

**Reply by matthias-mueller** [reply]
> If users provide a parameter like @Auth User user in their resource method, they rely on an authenticated user.

Not sure this expectation is covered by Jersey. I might be true for DW, which owns `@Auth`.

**Reply by zUniQueX** [reply]
> I was talking about _consistent injection behavior_ compared to the other annotations (not symmetry with potential producers inside DW)

Sorry, I then misunderstood this.

**Reply by zUniQueX** [reply]
> > If users provide a parameter like @Auth User user in their resource method, they rely on an authenticated user.
> 
> Not sure this expectation is covered by Jersey. I might be true for DW, which owns `@Auth`.

That's exactly my point. Personally I like the current approach. Additionally, it has been implemented this way for a long time, so we should have a good reason to change the current behavior. As the current implementation provides a way to fail without a principal (`@Auth User user`) and a way to not fail (`@Auth Optional<User> user`) I think it is more flexible and we should hand this degree of freedom to the users.

**Reply by matthias-mueller** [reply]
Ahh, now I can see where you are coming from and where you wish to put the guardrails. My reasoning was from the type system and I felt it natural to have an opt-in on using `Optional`s (or not).

So what you wish to preserve is an annotation / injection mechanism that fails with an `InternalServerError` if the Pricipal to be injected is NULL?

**Reply by zUniQueX** [reply]
That's correct. The current behavior is that the endpoint fails with an exception, if the principal _would_ be `null` and is not declared with an `Optional<T>` type. This provides the possibility to avoid a check, if the provided user is authenticated or not, as the auth mechanism would reject a request when there is no authenticated user.

**Reply by matthias-mueller** [reply]
But what is the case for throwing code `500` at the client? The only reason why I would opt-in to failing Principal injection is to prevent users accessing logic inside the method. But in that case the return value should be 401/403/404.

**Reply by zUniQueX** [reply]
Handling this is currently deferred to a user's implementation of an `AuthFilter`. The [corresponding logic](https://github.com/dropwizard/dropwizard/blob/release/2.1.x/dropwizard-auth/src/main/java/io/dropwizard/auth/basic/BasicCredentialAuthFilter.java#L27-L29) can be seen in the Dropwizard provided auth filters. The `IllegalStateException` only comes into play, if a user not correctly handles an unauthenticated request in the auth filter.

**Reply by matthias-mueller** [reply]
> The IllegalStateException only comes into play, if a user not correctly handles an unauthenticated request in the auth filter.

So what you mean is that any implementation of [AuthFilter](https://github.com/dropwizard/dropwizard/blob/70786c36271f047c971e603ce784d4752b48af12/dropwizard-auth/src/main/java/io/dropwizard/auth/AuthFilter.java) is expected to throw an exception if it cannot obtain a Principal? I assumed that simply doing nothing (e.g. if the user is unknown in a particular realm) would be a legal implementation.

**Reply by zUniQueX** [reply]
If no authentication information is provided, a status code 401 should be returned for which Dropwizard provides the `UnauthorizedHandler`. The application thereby shouldn't reach the resource, so this has to be done in the auth filter. Dropwizard doesn't enforce this to provide the users more flexibility in their implementations (e.g. not using an `UnauthorizedHandler` for this), but yes, this _should_ be done.

**Reply by matthias-mueller** [reply]
My understanding of the authentication control flow in DW has definitely improved over this discussion. The special distinction of `Optional` vs `nullable Principals` makes me still feel a bit uneasy, but there's sufficient tooling in DW to soothe my gut feeling. Thanks for your patience @zUniQueX !
